src/transport/TCP.h

```diff
@@ -0,0 +1,235 @@
+    {
+        PendingPacket emptyPending{
+            .peerAddress  = PeerAddress::Uninitialized(),
+            .packetBuffer = nullptr,
+        };
+
+        std::fill(mActiveConnections, mActiveConnections + mActiveConnectionsSize, nullptr);
+        std::fill(mPendingPackets, mPendingPackets + mPendingPacketsSize, emptyPending);
+    }
+    virtual ~TCPBase();
+
+    /**
+     * Initialize a TCP transport on a given port.
+     *
+     * @param params        TCP configuration parameters for this transport
+     *
+     * @details
+     *   Generally send and receive ports should be the same and equal to CHIP_PORT.
+     *   The class allows separate definitions to allow local execution of several
+     *   Nodes.
+     */
+    CHIP_ERROR Init(TcpListenParameters & params);
+
+    CHIP_ERROR SendMessage(const MessageHeader & header, const PeerAddress & address, System::PacketBuffer * msgBuf) override;
+
+    void Disconnect(const PeerAddress & address) override;
+
+    bool CanSendToPeer(const PeerAddress & address) override
+    {
+        return (mState == State::kInitialized) && (address.GetTransportType() == Type::kTcp) &&
+            (address.GetIPAddress().Type() == mEndpointType);
+    }
+
+    /**
+     * Helper method to determine if IO processing is still required for a TCP transport
+     * before everything is cleaned up (socket closing is async, so after calling 'Close' on
+     * the transport, some time may be needed to actually be able to close.)
+     */
+    bool HasActiveConnections() const;
+
+    /**
+     * Close all active connetions
+     */
+    void CloseActiveConnections();
+
+private:
+    /**
+     * Find an active connection to the given peer or return nullptr if
+     * no active connection exists.
+     */
+    Inet::TCPEndPoint * FindActiveConnection(const PeerAddress & addr);
+
+    /**
+     * Sends the specified message once a connection has been established.
+     *
+     * @param addr - what peer to connect to
+     * @param msg - what buffer to send once a connection has been established.
+     */
+    CHIP_ERROR SendAfterConnect(const PeerAddress & addr, System::PacketBuffer * msg);
+
+    /**
+     * Process a single received buffer from the specified peer address.
+     *
+     * @param endPoint the source end point from which the data comes from
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the actual data
+     */
+    CHIP_ERROR ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress, System::PacketBuffer * buffer);
+
+    /**
+     * Process a single message of the specified size from a buffer.
+     *
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the buffer containing the message
+     * @param messageSize how much of the head contains the actual message
+     */
+    CHIP_ERROR ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
```

Comment:
And here should document what the ownership semantics for buffer are, because I think that would make it clearer whether the caller and the implementation agree on them.

Reply:
I updated the code to correctly handle ownership. Also found the case where no handler was set was not clearning buffers.